tensorflow/core/kernels/bincount_op.cc

```diff
@@ -75,34 +66,71 @@ class BincountOp : public OpKernel {
           for (int64 i = start_ind; i < limit_ind; i++) {
             int32 value = arr(i);
             if (value < size) {
-              if (has_weights) {
-                partial_bins(worker_id, value) += weights(i);
-              } else {
-                // Complex numbers don't support "++".
-                partial_bins(worker_id, value) += T(1);
-              }
+              // Complex numbers don't support "++".
+              partial_bins(worker_id, value) += T(1);
             }
           }
         });
-    TensorShape output_shape({size});
-    Tensor* output_t;
-    OP_REQUIRES_OK(ctx, ctx->allocate_output(0, output_shape, &output_t));
+
     // Sum the partial bins along the 0th axis.
     Eigen::array<int, 1> reduce_dims({0});
-    output_t->flat<T>().device(ctx->eigen_cpu_device()) =
-        partial_bins.sum(reduce_dims);
+    output.device(context->eigen_cpu_device()) = partial_bins.sum(reduce_dims);
+    return Status::OK();
   }
 };
 
-#define REGISTER(TYPE)                                               \
+}  // namespace functor
+
+template <typename Device, typename T>
+class BincountOp : public OpKernel {
+ public:
+  explicit BincountOp(OpKernelConstruction* ctx) : OpKernel(ctx) {}
+
+  void Compute(OpKernelContext* ctx) override {
+    const Tensor& arr_t = ctx->input(0);
+    const Tensor& size_tensor = ctx->input(1);
+    const Tensor& weights_t = ctx->input(2);
+
+    int32 size = size_tensor.scalar<int32>()();
+    OP_REQUIRES(ctx, size >= 0, errors::InvalidArgument(
+                                    "size (", size, ") must be non-negative"));
+
+    OP_REQUIRES(
+        ctx, (weights_t.NumElements() == 0),
+        errors::InvalidArgument("Weights should not be passed as it should be "
```

Comment:
I'm sorry, I should've said this earlier, the CPU implementation needs to retain the ability to handle weights != None.  This is because of backwards compatibility - graphs that were created must still run correctly in a new version of TF.  (This is only true for the CPU version, as the GPU version didn't exist).

So basically, the CPU code needs to continue to be able to handle weights, but the graph code should stay as it is so that all _future_ graphs use unsorted_segment_sum instead.  Once six months have passed, we can remove the CPU code, although more likely, it will simply stay around forever.

Reply:
Thanks @ekelsen. The CPU implementation has been added back for weight != None.